util/java/libtiled-java/src/main/java/org/mapeditor/io/TMXMapReader.java

```diff
@@ -284,92 +270,90 @@ private TileSet unmarshalTileset(Node t, boolean isExternalTileset) throws Excep
                 return ext;
             }
         } else {
-            final int tileWidth = getAttribute(t, "tilewidth", map != null ? map.getTileWidth() : 0);
-            final int tileHeight = getAttribute(t, "tileheight", map != null ? map.getTileHeight() : 0);
-            final int tileSpacing = getAttribute(t, "spacing", 0);
-            final int tileMargin = getAttribute(t, "margin", 0);
-
             final String name = getAttributeValue(t, "name");
 
-            if (settings.reuseCachedTilesets) {
-                if (cachedTilesets == null) {
-                    cachedTilesets = new HashMap<>();
+            if (tilesetCache != null) {
+                if (tilesetCache.needToLoadTileset(name)) {
+                    TileSet tileSet = processTileset(t, name);
+                    tilesetCache.tilesetLoadingFinished(tileSet);
                 }
-
-                set = cachedTilesets.get(name);
-                if (set != null)
-                    return set;
-
-                set = new TileSet();
-                cachedTilesets.put(name, set);
-            } else {
-                set = new TileSet();
+                return tilesetCache.getTileset(name);
```

Comment:
Would it be reasonable to simplify this API so that the cache API is just `getTileset(name, lambda)`, so that it would call the lambda function to load the tileset when it does not exist in the cache yet? (I'm not sure what lambda functions would look like in Java, maybe:

```java
TileSet tileSet = tilesetCache.getTileset(name, (Node t, String name) -> processTileset(t, name))
```

(and the lambda doesn't really need to capture the name, since it could be read again from the `Node`)

Or are there some downsides to this?

Reply:
Or, I guess this should actually look more like:
```java
TileSet tileSet = tilesetCache.getTileset(name, () -> processTileset(t, name))
```
Since the `t` and `name` are local and can't be passed in again as parameters to the lambda. But I don't know how capturing of local values works in Java.